core/primitives/src/validator_mandates/compute_price.rs

```diff
@@ -0,0 +1,246 @@
+use {
+    super::ValidatorMandatesConfig,
+    near_primitives_core::types::Balance,
+    std::cmp::{min, Ordering},
+};
+
+/// Given the stakes for the validators and the target number of mandates to have,
+/// this function computes the mandate price to use. It works by iterating a
+/// function in an attempt to find its fixed point. This function is motived as follows:
+/// Let the validator stakes be denoted by `s_i` a let `S = \sum_i s_i` be the total
+/// stake. For a given mandate price `m` we can write each `s_i = m * q_i + r_i`
+/// (by the Euclidean algorithm). Hence, the number of whole mandates created by
+/// that price is equal to `\sum_i q_i`. If we set this number of whole mandates
+/// equal to the target number `N` then substitute back in to the previous equations
+/// we have `S = m * N + \sum_i r_i`. We can rearrange this to solve for `m`,
+/// `m = (S - \sum_i r_i) / N`. Note that `r_i = a_i % m` so `m` is not truly
+/// isolated, but rather the RHS is the expression we want to find the fixed point for.
+pub fn compute_mandate_price<F, I>(config: ValidatorMandatesConfig, stakes: F) -> Balance
+where
+    I: Iterator<Item = Balance>,
+    F: Fn() -> I,
+{
+    let ValidatorMandatesConfig { target_mandates_per_shard, num_shards } = config;
+    let total_stake = saturating_sum(stakes());
+
+    // The target number of mandates cannot be larger than the total amount of stake.
+    // In production the total stake is _much_ higher than
+    // `num_shards * target_mandates_per_shard`, but in tests validators are given
+    // low staked numbers, so we need to have this condition in place.
+    let target_mandates: u128 =
+        min(num_shards.saturating_mul(target_mandates_per_shard) as u128, total_stake);
+
+    let initial_price = total_stake / target_mandates;
+
+    // Function to compute the new estimated mandate price as well as
+    // evaluate the given mandate price.
+    let f = |price: u128| {
+        let mut whole_mandates = 0_u128;
+        let mut remainders = 0_u128;
+        for s in stakes() {
+            whole_mandates = whole_mandates.saturating_add(s / price);
+            remainders = remainders.saturating_add(s % price);
+        }
+        let updated_price = if total_stake > remainders {
+            (total_stake - remainders) / target_mandates
+        } else {
```

Comment:
It is always true that `total_stake >= remainders`, right?

Reply:
Not relevant after https://github.com/near/nearcore/pull/11044/commits/5ca2a4864f0e9bbd84fc1859893972a56d5ea5d8